File: src/ch/jbead/BeadUndo.java
/** jbead - http://www.jbead.ch
    Copyright (C) 2001-2012  Damian Brunold

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package ch.jbead;

/**
 * 
 */
public class BeadUndo {

    private static final int MAXUNDO = 100;

    private BeadField[] data = new BeadField[MAXUNDO];
    private boolean modified[] = new boolean[MAXUNDO];
    private int first;
    private int last;
    private int current;

    public BeadUndo() {
        first = 0;
        last = 0;
        current = 0;
        for (int i = 0; i < MAXUNDO; i++) {
            data[i] = new BeadField();
            modified[i] = false;
        }
    }

    public boolean canUndo() {
        return current != first;
    }

    public boolean canRedo() {
        return current != last;
    }

    public void clear() {
        first = 0;
        last = 0;
        current = 0;
    }

    public void snapshot(BeadField data, boolean modified) {
        this.data[current].copyFrom(data);
        this.modified[current] = modified;
        current = (current + 1) % MAXUNDO;
        if (current == first) first = (first + 1) % MAXUNDO;
        last = current;
    }

    public void prepareSnapshot(BeadField data, boolean modified) {
        if (!modified) return;
        this.data[current].copyFrom(data);
        this.modified[current] = modified;
    }

    public void undo(BeadField data) {
        if (current == first) return;
        current = (current - 1 + MAXUNDO) % MAXUNDO;
        data.copyFrom(this.data[current]);
    }

    public void redo(BeadField data) {
        if (current == last) return;
        current = (current + 1) % MAXUNDO;
        data.copyFrom(this.data[current]);
    }

    public boolean isModified() {
        return modified[current];
    }

}

File: src/domain/beadundo.cpp
#include "beadundo.h"

namespace jbead {

BeadUndo::BeadUndo() = default;

void BeadUndo::clear()
{
    m_first = m_last = m_current = 0;
}

void BeadUndo::snapshot(const BeadField& data, const QList<QColor>& colors,
                        const QString& symbols, bool modified)
{
    m_data[m_current].copyFrom(data);
    m_colors[m_current]   = colors;
    m_symbols[m_current]  = symbols;
    m_modified[m_current] = modified;
    m_current = (m_current + 1) % MAX_UNDO;
    if (m_current == m_first) m_first = (m_first + 1) % MAX_UNDO;
    m_last = m_current;
}

void BeadUndo::prepareSnapshot(const BeadField& data, const QList<QColor>& colors,
                               const QString& symbols, bool modified)
{
    if (!modified) return;
    m_data[m_current].copyFrom(data);
    m_colors[m_current]   = colors;
    m_symbols[m_current]  = symbols;
    m_modified[m_current] = modified;
}

void BeadUndo::undo(BeadField& data, QList<QColor>& colors, QString& symbols)
{
    if (m_current == m_first) return;
    m_current = (m_current - 1 + MAX_UNDO) % MAX_UNDO;
    data.copyFrom(m_data[m_current]);
    colors  = m_colors[m_current];
    symbols = m_symbols[m_current];
}

void BeadUndo::redo(BeadField& data, QList<QColor>& colors, QString& symbols)
{
    if (m_current == m_last) return;
    m_current = (m_current + 1) % MAX_UNDO;
    data.copyFrom(m_data[m_current]);
    colors  = m_colors[m_current];
    symbols = m_symbols[m_current];
}

} // namespace jbead
